src/mpid/ch4/shm/posix/release_gather/release_gather.c

```diff
@@ -314,7 +314,7 @@ int MPIDI_POSIX_mpi_release_gather_comm_init(MPIR_Comm * comm_ptr,
         if (MPIR_CVAR_ENABLE_INTRANODE_TOPOLOGY_AWARE_TREES &&
             getenv("HYDRA_USER_PROVIDED_BINDING")) {
             /* Topology aware trees are created only when the user has specified process binding */
-            if (hwloc_topology_load(MPIR_Process.hwloc_topology) == 0) {
+            if (MPIR_Process.hwloc_topology != NULL) {
```

Comment:
@jain-surabhi-23 there is a `MPIR_Process.bindset_is_valid` variable that is used to signal `hwloc` is available and `bindset` is valid. This is what is also used in `comm_split_type` code: https://github.com/pmodels/mpich/blob/d1675db67908601ee304b8f6d227c414160dfec2/src/mpi/comm/comm_split_type.c#L868

Reply:
I have tried it. It does not work for me. Because after `finalize` is already called, `hwloc_topology` is destroyed but it is not reflected through the `bindset_is_valid`. Hence, topology aware trees are attempted to create, which results in segmentation fault.

Reply:
Sorry I don't think I understand. Why is it important that `bindset_is_valid` is not reset to zero when library is finalized?

Reply:
When library is in the process of being finalized, `hwloc_topology` is already destroyed, and some `Bcast`, `Reduce` or `Allreduce` calls are internally made. This reaches the shared memory collectives and tries to create topology aware trees. But we want to fallback to non-topology aware trees because `hwloc_topology` is no longer valid. For this to happen `bindset_is_valid` needs to reset as soon as `hwloc_topology` is destroyed.

Reply:
So we can either keep my change. Or I can use `bindset_is_valid`, but I will have to explicitly set it to `0` after `hwloc_topology_destroy(MPIR_Process.hwloc_topology);` in finalize.c

Reply:
If `hwloc` is finalized too early that is a bug. I would move the finalization of `hwloc` topology as very last step and allow the library to keep using topology aware collectives as normal.

Reply:
Yeah. That is one thing I can look into. Other bug is `bindset_is_valid` is not reflecting the fact that `hwloc_topology` is destroyed.